File: server/m29-physical-evidence/mask.ts
import type { BBox } from "../../shared/types";
import { colorDistance, rgbAt } from "./image";
import type { BackgroundEstimate, DecodedRgbaImage, ForegroundMask, Mask } from "./types";

export function newMask(width: number, height: number): Mask {
  return { width, height, data: new Uint8Array(width * height) };
}

export function inBounds(mask: Mask, x: number, y: number): boolean {
  return x >= 0 && y >= 0 && x < mask.width && y < mask.height;
}

export function getMask(mask: Mask, x: number, y: number): boolean {
  if (!inBounds(mask, x, y)) return false;
  return mask.data[y * mask.width + x] === 1;
}

export function setMask(mask: Mask, x: number, y: number, value: boolean): void {
  if (!inBounds(mask, x, y)) return;
  mask.data[y * mask.width + x] = value ? 1 : 0;
}

export function fillBBox(mask: Mask, bbox: BBox, padding: number): void {
  const x1 = Math.max(0, bbox.x - padding);
  const y1 = Math.max(0, bbox.y - padding);
  const x2 = Math.min(mask.width, bbox.x + bbox.width + padding);
  const y2 = Math.min(mask.height, bbox.y + bbox.height + padding);
  for (let y = y1; y < y2; y += 1) {
    for (let x = x1; x < x2; x += 1) {
      setMask(mask, x, y, true);
    }
  }
}

export function bboxMask(width: number, height: number, bbox: BBox): Mask {
  const mask = newMask(width, height);
  fillBBox(mask, bbox, 0);
  return mask;
}

export function createForegroundMask(image: DecodedRgbaImage, background: BackgroundEstimate, textMask = newMask(image.width, image.height)): ForegroundMask {
  const mask = newMask(image.width, image.height) as ForegroundMask;
  let foregroundPixelCount = 0;
  for (let y = 0; y < image.height; y += 1) {
    for (let x = 0; x < image.width; x += 1) {
      if (getMask(textMask, x, y)) continue;
      if (colorDistance(rgbAt(image, x, y), background.color) > background.threshold) {
        setMask(mask, x, y, true);
        foregroundPixelCount += 1;
      }
    }
  }
  mask.foregroundPixelCount = foregroundPixelCount;
  return mask;
}

File: archive/legacy-code/services/backend-go/internal/m29/mask/mask.go
package mask

import "github.com/luqing-studio/image-figma/services/backend-go/internal/m29/contract"

type Mask struct {
	Width  int
	Height int
	Data   []bool
}

func New(width, height int) Mask {
	return Mask{Width: width, Height: height, Data: make([]bool, width*height)}
}

func (m Mask) InBounds(x, y int) bool {
	return x >= 0 && y >= 0 && x < m.Width && y < m.Height
}

func (m Mask) Get(x, y int) bool {
	if !m.InBounds(x, y) {
		return false
	}
	return m.Data[y*m.Width+x]
}

func (m Mask) Set(x, y int, value bool) {
	if m.InBounds(x, y) {
		m.Data[y*m.Width+x] = value
	}
}

func (m Mask) Count() int {
	total := 0
	for _, value := range m.Data {
		if value {
			total++
		}
	}
	return total
}

func FillBBox(m Mask, bbox contract.BBox, padding int) {
	x1 := max(0, bbox.X-padding)
	y1 := max(0, bbox.Y-padding)
	x2 := min(m.Width, bbox.X+bbox.Width+padding)
	y2 := min(m.Height, bbox.Y+bbox.Height+padding)
	for y := y1; y < y2; y++ {
		for x := x1; x < x2; x++ {
			m.Set(x, y, true)
		}
	}
}

func BBoxMask(width, height int, bbox contract.BBox) Mask {
	m := New(width, height)
	FillBBox(m, bbox, 0)
	return m
}
